src/backend/routers/files.ts

```diff
@@ -0,0 +1,90 @@
+import { z } from "zod";
+import {
+  PutObjectCommand,
+  HeadObjectCommand,
+  GetObjectCommand,
+} from "@aws-sdk/client-s3";
+import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
+import { protectedProcedure } from "../trpc";
+import { randomUUID } from "crypto";
+
+export const fileProcedures = {
+  getAllFiles: protectedProcedure.query(async (req) => {
+    return req.ctx.db.selectFrom("file").selectAll().execute();
+  }),
+
+  getPresignedUrlForFileDownload: protectedProcedure
+    .input(
+      z.object({
+        file_id: z.string().uuid(),
+      })
+    )
+    .mutation(async (req) => {
+      const file = await req.ctx.db
+        .selectFrom("file")
+        .select("ext_s3_path")
+        .where("file_id", "=", req.input.file_id)
+        // Only allow access to your own files
+        .where("uploaded_by_user_id", "=", req.ctx.auth.userId)
+        .executeTakeFirstOrThrow();
+
+      const command = new GetObjectCommand({
+        Bucket: req.ctx.env.S3_BUCKET_NAME,
+        Key: file.ext_s3_path,
+      });
+
+      const url = await getSignedUrl(req.ctx.s3, command, {
+        expiresIn: 60 * 60,
+      });
+      return url;
+    }),
+
+  getPresignedUrlForFileUpload: protectedProcedure
+    .input(
+      z.object({
+        type: z.string(),
+      })
+    )
+    .mutation(async (req) => {
+      const key = randomUUID();
+
+      const command = new PutObjectCommand({
+        Bucket: req.ctx.env.S3_BUCKET_NAME,
```

Comment:
IIUC, the path of the file will be s3://bucket/key, where bucket="compass" and key=uuid. What is the best practice for working with S3? Have multiple buckets for an app, e.g. change the bucket to say "compass_files" or have one parent bucket with sub-directories, e.g. "compass/files/..."?

Reply:
I haven't worked with buckets much but I think it makes sense to prefer multiple buckets over folders, as you can set different access controls on different buckets but it's harder to/impossible at the directory level